File: src/commands/category.ts
import type { Category, CategoryWithChildren, NewCategory, UpdateCategory } from "@/lib/types";
import { getRunEnv, RUN_ENV, invokeTauri, invokeWeb } from "@/adapters";
import { logger } from "@/adapters";

export const getCategories = async (): Promise<Category[]> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("get_categories");
      case RUN_ENV.WEB:
        return invokeWeb("get_categories");
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error fetching categories.");
    throw error;
  }
};

export const getCategoriesHierarchical = async (): Promise<CategoryWithChildren[]> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("get_categories_hierarchical");
      case RUN_ENV.WEB:
        return invokeWeb("get_categories_hierarchical");
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error fetching hierarchical categories.");
    throw error;
  }
};

export const getExpenseCategories = async (): Promise<CategoryWithChildren[]> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("get_expense_categories");
      case RUN_ENV.WEB:
        return invokeWeb("get_expense_categories");
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error fetching expense categories.");
    throw error;
  }
};

export const getIncomeCategories = async (): Promise<CategoryWithChildren[]> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("get_income_categories");
      case RUN_ENV.WEB:
        return invokeWeb("get_income_categories");
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error fetching income categories.");
    throw error;
  }
};

export const createCategory = async (category: NewCategory): Promise<Category> => {
  try {
    // Ensure isIncome is a proper boolean (not 0/1 from database)
    const normalizedCategory = {
      ...category,
      isIncome: !!category.isIncome, // Double negation ensures true boolean
    };

    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("create_category", { category: normalizedCategory });
      case RUN_ENV.WEB:
        return invokeWeb("create_category", { category: normalizedCategory });
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error creating category.");
    throw error;
  }
};

export const updateCategory = async (id: string, update: UpdateCategory): Promise<Category> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("update_category", { id, update });
      case RUN_ENV.WEB:
        return invokeWeb("update_category", { id, update });
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error updating category.");
    throw error;
  }
};

export const deleteCategory = async (categoryId: string): Promise<void> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        await invokeTauri("delete_category", { categoryId });
        return;
      case RUN_ENV.WEB:
        await invokeWeb("delete_category", { categoryId });
        return;
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error deleting category.");
    throw error;
  }
};

File: src-tauri/src/commands/category.rs
use std::collections::HashMap;
use std::sync::Arc;

use crate::context::ServiceContext;
use log::{debug, error};
use tauri::State;

use wealthfolio_core::categories::{Category, CategoryWithChildren};

#[derive(serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateCategoryInput {
    pub name: String,
    pub parent_id: Option<String>,
    pub color: Option<String>,
    pub icon: Option<String>,
    pub is_income: bool,
}

#[derive(serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateCategoryInput {
    pub name: Option<String>,
    pub color: Option<String>,
    pub icon: Option<String>,
    pub sort_order: Option<i32>,
}

#[tauri::command]
pub async fn get_categories(
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<Category>, String> {
    debug!("Fetching all categories...");
    state
        .category_service()
        .get_all_categories()
        .map_err(|e| {
            error!("Failed to fetch categories: {}", e);
            format!("Failed to fetch categories: {}", e)
        })
}

#[tauri::command]
pub async fn get_categories_hierarchical(
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<CategoryWithChildren>, String> {
    debug!("Fetching hierarchical categories...");
    state
        .category_service()
        .get_categories_hierarchical()
        .map_err(|e| {
            error!("Failed to fetch hierarchical categories: {}", e);
            format!("Failed to fetch hierarchical categories: {}", e)
        })
}

#[tauri::command]
pub async fn get_expense_categories(
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<CategoryWithChildren>, String> {
    debug!("Fetching expense categories...");
    state
        .category_service()
        .get_expense_categories()
        .map_err(|e| {
            error!("Failed to fetch expense categories: {}", e);
            format!("Failed to fetch expense categories: {}", e)
        })
}

#[tauri::command]
pub async fn get_income_categories(
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<CategoryWithChildren>, String> {
    debug!("Fetching income categories...");
    state
        .category_service()
        .get_income_categories()
        .map_err(|e| {
            error!("Failed to fetch income categories: {}", e);
            format!("Failed to fetch income categories: {}", e)
        })
}

#[tauri::command]
pub async fn create_category(
    category: CreateCategoryInput,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Category, String> {
    debug!("Creating category: {:?}", category.name);
    state
        .category_service()
        .create_category(
            category.name,
            category.parent_id,
            category.color,
            category.icon,
            category.is_income,
        )
        .await
        .map_err(|e| {
            error!("Failed to create category: {}", e);
            format!("Failed to create category: {}", e)
        })
}

#[tauri::command]
pub async fn update_category(
    id: String,
    update: UpdateCategoryInput,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Category, String> {
    debug!("Updating category: {}", id);
    state
        .category_service()
        .update_category(
            &id,
            update.name,
            update.color,
            update.icon,
            update.sort_order,
        )
        .await
        .map_err(|e| {
            error!("Failed to update category: {}", e);
            format!("Failed to update category: {}", e)
        })
}

#[tauri::command]
pub async fn delete_category(
    category_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<(), String> {
    debug!("Deleting category: {}", category_id);
    state
        .category_service()
        .delete_category(&category_id)
        .await
        .map_err(|e| {
            error!("Failed to delete category: {}", e);
            format!("Failed to delete category: {}", e)
        })?;
    Ok(())
}

#[tauri::command]
pub async fn get_category_activity_counts(
    state: State<'_, Arc<ServiceContext>>,
) -> Result<HashMap<String, i64>, String> {
    debug!("Fetching category activity counts...");
    state
        .category_service()
        .get_activity_counts()
        .map_err(|e| {
            error!("Failed to fetch category activity counts: {}", e);
            format!("Failed to fetch category activity counts: {}", e)
        })
}
